### PR TITLE
Update multi-chain permutation and permutation unittest 

### PR DESCRIPTION
**Pull request by dingquanyu**
Hi @christinaflo  and @jnwei 

Sorry I forgot to update the unittest for multi-chain permutations after the major updates on the functions. Here I have added necessary steps to prepare fake test data for testing these functions.  Now all the 3 tests can run successfully. Hope it helps.

BTW I'm now adding typing to the functions in multi_chain_permutation.py and fixing some comments in that file as well. These can however go to another PR if you prefer?

Dingquan

### CODE REVIEW

**Review by jnwei — CHANGES_REQUESTED** [review]
Thank you so much for expanding the docstrings in multi_chain_permutation.py and for fixing the tests. 

There is a lot of non-trivial code associated with determining the best permutations, and the docstrings will go a long way towards making the code more approachable.

I have some suggestions to help further improve the clarity of the code, but overall, really nice work!

**Comment by dingquanyu** [comment]
@jnwei Many thanks for your suggestions and reviews :D  I've updated the PR in the new commit

File: openfold/utils/multi_chain_permutation.py [hunk 1]
```diff
@@ -21,19 +32,19 @@ def compute_rmsd(
     return torch.sqrt(msd + eps)  # prevent sqrt 0
 
 
-def kabsch_rotation(P, Q):
+def kabsch_rotation(P:torch.Tensor, Q:torch.Tensor) -> torch.Tensor:
```
Comment (jnwei) [inline openfold/utils/multi_chain_permutation.py:35]: nit: Please add spaces between arguments and type definition.

File: openfold/utils/multi_chain_permutation.py [hunk 2]
```diff
@@ -281,14 +346,16 @@ def get_entity_2_asym_list(batch):
     return entity_2_asym_list
 
 
-def calculate_input_mask(true_ca_masks, anchor_gt_idx, anchor_gt_residue,
-                         asym_mask, pred_ca_mask):
+def calculate_input_mask(true_ca_masks: List[torch.Tensor], anchor_gt_idx: torch.Tensor, 
+                         anchor_gt_residue: list,
```
Comment (jnwei) [inline openfold/utils/multi_chain_permutation.py:350]: nit: The docstring the type is a Tensor, is this a list or a Tensor?

File: openfold/utils/multi_chain_permutation.py [hunk 3]
```diff
@@ -303,11 +370,38 @@ def calculate_input_mask(true_ca_masks, anchor_gt_idx, anchor_gt_residue,
     return input_mask
 
 
-def calculate_optimal_transform(true_ca_poses,
-                                anchor_gt_idx, anchor_gt_residue,
-                                true_ca_masks, pred_ca_mask,
-                                asym_mask,
-                                pred_ca_pos):
+def calculate_optimal_transform(true_ca_poses: List[torch.Tensor],
+                                anchor_gt_idx: int, anchor_gt_residue: list,
```
Comment (jnwei) [inline openfold/utils/multi_chain_permutation.py:374]: Same thing here, is anchor_gt_residue a list or a tensor?

File: tests/test_permutation.py [hunk 4]
```diff
@@ -155,15 +156,30 @@ def test_3_merge_labels(self):
...
+            'aatype': torch.randint(21, size=(1, 325)),
+            'seq_length': torch.tensor([57])
+        }
+        fake_input_features['asym_id'] = fake_input_features['asym_id'].reshape(1, 325)
+        fake_input_features["residue_index"] = pad_features(torch.tensor(self.residue_index).reshape(1, 57), nres_pad, pad_dim=1)
+        fake_input_features['all_atom_positions'] = pad_features(true_atom_position, nres_pad, pad_dim=1)
+        fake_input_features['all_atom_mask'] = pad_features(true_atom_mask, nres_pad=nres_pad, pad_dim=1)
+        
+        # NOTE
+        # batch: simulates ground_truth features
```
Comment (jnwei) [inline tests/test_permutation.py:177]: nit: replace gonna with going to

File: tests/test_permutation.py [hunk 5]
```diff
@@ -111,26 +113,25 @@ def test_2_permutation_pentamer(self):
         batch['all_atom_positions'] = true_atom_position
         batch['all_atom_mask'] = true_atom_mask
 
-        aligns, _ = compute_permutation_alignment(out, batch,
+        aligns, per_asym_residue_index = compute_permutation_alignment(out, batch,
                                                   batch)
-        print(f"##### aligns is {aligns}")
         possible_outcome = [[(0, 1), (1, 0), (2, 3), (3, 4), (4, 2)], [(0, 0), (1, 1), (2, 3), (3, 4), (4, 2)]]
```
[the remark this answers is not in the document]
Comment (jnwei) [reply tests/test_permutation.py:118]: Just a reminder here, a comment explaining why you expect the given possible outcome, and why the wrong_outcome is bad would be very helpful.

To help explain the examples, you could even break up the examples into different variables with acceptable / not acceptable cases. For example:

```
chain_a_permuted = [(0, 1), (1, 0), (2, 2), (3, 3), (4, 4)]
chain_b_permuted = [(0, 0), (1, 1), (2, 3), (3, 4), (4, 2)]
chains_a_and_b_permuted = [(0, 1), (1, 0), (2, 3), (3, 4), (4, 2)]
no_permutation = [(0, 0), (1, 1), (2, 2), (3, 3), (4, 4)]

possible_outcome = [chain_a_permuted, chain_b_permuted]
wrong_outcome = [chain_a_and_b_permuted, no_permutation] 
```

Although in this example, I still don't understand why `chain_a_and_b_permuted` would be under `wrong_outcome`

**Review by jnwei — CHANGES_REQUESTED** [review]
Thanks for updating the docstrings Dingquan! Just a few more minor comments

**Review by jnwei — APPROVED** [review]

**Comment by jnwei** [comment]
Thanks for the additions to the docstring Dingquan! The explanation for the tests are much clearer now.

**Pull request merged**